### PR TITLE
Validate and pick consecutive txs in L1 consensus

### PR DESCRIPTION
**Pull request by TheMMaciek**

### CODE REVIEW

File: modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala [hunk 1]
```diff
@@ -2,33 +2,56 @@ package org.tessellation.dag.transaction.filter
...
 import scala.annotation.tailrec
 
 import org.tessellation.kryo.KryoSerializer
 import org.tessellation.schema.transaction.{Transaction, TransactionReference}
 import org.tessellation.security.Hashed
+import org.tessellation.security.signature.Signed
 
 object Consecutive {
+  val hashedTxOrder: Order[Hashed[Transaction]] =
+    Order.whenEqual(Order.by(-_.fee.value.value), Order[Hashed[Transaction]])
```
Comment (kpudlik) [inline modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala:20]: You can achieve more DRY solution here.
`Hashed[Transaction]` has already `Signed[Transaction]` inside so you can create `Order[Hashed[Transaction]]` by using contravariant functor:

```scala
import cats.syntax.contravariant._

val hashedTxOrder: Order[Hashed[Transaction]] = signedTxOrder.contramap(_.signed)
```

**Review by kpudlik — CHANGES_REQUESTED** [review]

File: modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/consensus/block/RoundData.scala [hunk 2]
```diff
@@ -44,9 +54,22 @@ case class RoundData(
...
-  def formBlock(): Option[DAGBlock] =
-    NonEmptySet
-      .fromSet((ownProposal.transactions ++ peerProposals.values.flatMap(_.transactions).toSet).toSortedSet)
-      .map(DAGBlock(tips.value, _))
+  def formBlock[F[_]: Async: KryoSerializer: Logger](validator: TransactionValidator[F]): F[Option[DAGBlock]] =
+    (ownProposal.transactions ++ peerProposals.values.flatMap(_.transactions)).toList
+      .traverse(validator.validate)
+      .flatMap { validatedTxs =>
+        val invalid = validatedTxs.collect { case Invalid(e) => e }
+        val valid = validatedTxs.collect { case Valid(tx) => tx }
```
Comment (kriscieslak) [inline modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/consensus/block/RoundData.scala:63]: What about this: `val (invalid, valid) = validatedTxs.partitionMap(_.toEither)` ?

File: modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala [hunk 3]
```diff
@@ -2,33 +2,56 @@ package org.tessellation.dag.transaction.filter
...
+    takeGeneric[Id, TransactionReference, Hashed[Transaction]](
+      hashedTx => Id(TransactionReference.of(hashedTx)),
       _.parent,
       txs,
       lastAcceptedTxRef
-    )(Applicative[F], Order[TransactionReference], hashedTxOrder)
-  }
+    )(Applicative[Id], Order[TransactionReference], hashedTxOrder)
 
   private def takeGeneric[F[_]: Applicative, Ref: Order, Item: Order](
```
Comment (kriscieslak) [inline modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala:56]: It is possible here to substitute `Ref` with `TransactionReference`:

```
private def takeGeneric[F[_]: Applicative, Item: Order](
    refOf: Item => F[TransactionReference],
    prevRefOf: Item => TransactionReference,
    items: List[Item],
    initRef: TransactionReference
  )
```

File: modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala (hunk 1 above)
Comment (tbekas) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala:20]: Is it necessary though? `Hashed` has auto unwrap. This should work

```scala
val hashedTxOrder: Order[Hashed[Transaction]] = Order.reverse(Order.by(_.fee))
```

File: modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala (hunk 3 above)
Comment (tbekas) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala:56]: `Ref` is a parameter and `TransactionReference` is an argument to that parameter, so it already is the way you describe.

**Review by tbekas — DISMISSED** [review]
LGTM, but please use the `partitionMap` that @kriscieslak mentioned.

**Comment by kpudlik** [comment]
Please add the label :)

File: modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala (hunk 1 above)
Comment (TheMMaciek) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala:20]: We not only need it to sort by fee but also to be deterministic when different nodes run the same selection process. So I need to order by fee and then by everything else...

File: modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala (hunk 3 above)
Comment (kriscieslak) [reply modules/dag-shared/src/main/scala/org/tessellation/dag/transaction/filter/Consecutive.scala:56]: I just thought we can avoid explicitly passing `Ref: Order` this way, but I agree there is not much profit from this.

**Review by kriscieslak — APPROVED** [review]

**Review by kpudlik — APPROVED** [review]

**Pull request merged**